src/ck-core/ckarray.C

```diff
@@ -1471,21 +1471,20 @@ void CkArray::recvBroadcast(CkMessage *m)
     // extract this field here so we can still check it even if msg is freed
     const auto zc_msgtype = CMI_ZC_MSGTYPE(UsrToEnv(msg));
 
-    // All done, deliver to the first element
-    if (zc_msgtype == CMK_ZC_BCAST_RECV_ALL_DONE_MSG) {
+    unsigned int firstElemIndex=0; // used for delivery only to the first local element
 
-      unsigned int i=0;
-      bool doFree = true;
-      if (stableLocations && i == len-1) doFree = true;
-      broadcaster->deliver(msg, (ArrayElement*)localElemVec[i], doFree);
+    if (zc_msgtype == CMK_ZC_BCAST_RECV_ALL_DONE_MSG) { // message contains pointers to the posted buffer, which contains the data received
+      // All operations done, already consumed by other array elements, now deliver to the first element
 
-    } else if (zc_msgtype == CMK_ZC_BCAST_RECV_MSG) { // deliver to all array elements
+      bool doFree = true; // free it since all ops are done
+      broadcaster->deliver(msg, (ArrayElement*)localElemVec[firstElemIndex], doFree);
 
+    } else if (zc_msgtype == CMK_ZC_BCAST_RECV_MSG) { // message is used by the receiver to post the receiver buffer
+      // Initial metadata message, send only to the first element, other elements are sent CMK_ZC_BCAST_RECV_DONE_MSG after rget completion
 
-      unsigned int i=0;
-      bool doFree = false;
-      if (stableLocations && i == len-1) doFree = true;
-      broadcaster->deliver(msg, (ArrayElement*)localElemVec[i], doFree);
+      bool doFree = false; // do not free since msg will be reused to send buffers to peers,
```

Comment:
Previously there was a chance here that doFree was set to true (if stableLocations and len==1 right?). I'm not sure if we either need to keep that, or if it was a bug.

Reply:
Yeah, it shouldn't be freed since the same message is re-used and freed later. I'll run a test with len=1 and ensure that it crashes with the previous code.

Reply:
The previous code never crashed because `stableLocations` was false and that caused `doFree` to never be true. But that was a bug nevertheless since the message shouldn't be freed.